### PR TITLE
Add login success attribute and login failed event

### PR DESCRIPTION
**Pull request by AnjanaSamindraPerera**
## Related PR

[https://github.com/wso2-enterprise/asgardeo-extensions/pull/229](https://www.google.com/url?q=https://github.com/wso2-enterprise/asgardeo-extensions/pull/229&sa=D&source=calendar&ust=1707041976587724&usg=AOvVaw00zF_ZvSLDj9MZD9BVeMaM)

## Related Issue

https://github.com/wso2-enterprise/iam-engineering/issues/963

### CODE REVIEW

File: asyncapi/asgardeo/asyncapi.yml [hunk 1]
```diff
@@ -230,6 +234,52 @@ components:
...
+          example: myorg
+        userId:
+          type: string
+          example: e3dca90f-e1e3-4f4b-83e4-5743ecce570f
+        authenticatingUser:
+          type: string
+          example: john@gmail.com
+        serviceProvider:
+          type: string
+          example: My Account
```
Comment (RolandHewage) [inline asyncapi/asgardeo/asyncapi.yml:270]: Add descriptions for each of these properties

File: asyncapi/asgardeo/data_types.bal [hunk 2]
```diff
@@ -92,6 +92,15 @@ public type LoginSuccessEvent record {
     GenericSecurityData securityData?;
 };
 
+# Login Failed event
+#
+# + eventData - Event data  
+# + securityData - Event related security data
+public type LoginFailedEvent record {
+    LoginFailedData eventData?;
+    GenericSecurityData securityData?;
```
Comment (RolandHewage) [inline asyncapi/asgardeo/data_types.bal:101]: Seems these fields are optional. Make sure they aren't nillable. If they are nillable please handle it too

File: asyncapi/asgardeo/asyncapi.yml [hunk 3]
```diff
@@ -230,6 +234,52 @@ components:
         serviceProvider:
           type: string
           example: My Account
+        authSteps:
+          type: array
+          description: A list of authentication steps
+          items:
+            $ref: '#/components/schemas/authSteps'
```
Comment (RolandHewage) [inline asyncapi/asgardeo/asyncapi.yml:241]: We have referred the schema `authSteps` here. But we have defined a different schema `authStep`.

File: asyncapi/asgardeo/data_types.bal [hunk 4]
```diff
@@ -141,6 +151,26 @@ public type LoginSuccessData record {
...
+# + userId - User Id
+# + failedStep -  A list of Authentication Steps
+public type LoginFailedData record {
+    int organizationId?;
+    string ref?;
+    string organizationName?;
+    string serviceProvider?;
+    string authenticatingUser?;
+    string userId?;
+    AuthStep authSteps?;
```
Comment (RolandHewage) [inline asyncapi/asgardeo/data_types.bal:173]: We have defined `failedStep` in asyncapi.yaml. But here we have added `authSteps`. Please recheck

File: asyncapi/asgardeo/service_types.bal [hunk 5]
```diff
@@ -14,6 +14,7 @@ public type UserOperationService service object {
 
 public type LoginService service object {
     remote function onLoginSuccess(LoginSuccessEvent event) returns error?;
+    remote function onLoginFailed(LoginFailedEvent event) returns error?;
```
Comment (RolandHewage) [inline asyncapi/asgardeo/service_types.bal:17]: Add a comment quite similar to https://github.com/ballerina-platform/asyncapi-triggers/blob/ead5cbc9011c7d56af17761e4d45f8b5cb2f78d0/asyncapi/shopify/service_types.bal#L46

**Review by RolandHewage — COMMENTED** [review]
Please address the above changes.
Please update the Module.md with the new event details.
Please verify with @Drifftr whether this is the correct branch to do the release.

File: asyncapi/asgardeo/service_types.bal (hunk 5 above)
Comment (AnjanaSamindraPerera) [reply asyncapi/asgardeo/service_types.bal:19]: Resoloved in [e8d6609](https://github.com/ballerina-platform/asyncapi-triggers/pull/154/commits/e8d660936bbc0a411944c843eb7e01bbc72d117c)

File: asyncapi/asgardeo/data_types.bal (hunk 4 above)
Comment (AnjanaSamindraPerera) [reply asyncapi/asgardeo/data_types.bal]: Resolved in [e8d6609](https://github.com/ballerina-platform/asyncapi-triggers/pull/154/commits/e8d660936bbc0a411944c843eb7e01bbc72d117c)

File: asyncapi/asgardeo/asyncapi.yml (hunk 3 above)
Comment (AnjanaSamindraPerera) [reply asyncapi/asgardeo/asyncapi.yml]: Resolved in [e8d6609](https://github.com/ballerina-platform/asyncapi-triggers/pull/154/commits/e8d660936bbc0a411944c843eb7e01bbc72d117c)

**Review by Drifftr — APPROVED** [review]

**Review by NipunaRanasinghe — APPROVED** [review]
Approved as per https://github.com/ballerina-platform/asyncapi-triggers/pull/158

**Pull request merged**